packages/flutter/lib/src/material/tabs.dart

```diff
@@ -760,543 +352,478 @@ class TabBar<T> extends Scrollable implements AppBarBottomWidget {
-      _updateItemsFromChildren(0, 1);
-    } else if (selectedIndex == _tabCount - 1) {
-      _updateItemsFromChildren(selectedIndex - 1, selectedIndex);
-    } else {
-      _updateItemsFromChildren(selectedIndex - 1, selectedIndex, selectedIndex + 1);
-    }
+  TabController _controller;
+  List<Widget> _children;
+  double _offsetAnchor;
+  double _offsetBias = 0.0;
+  int _currentIndex;
+  bool _warpUnderway = false;
+
+  void _initTabController() {
+    if (_controller != null)
+      _controller.animation.removeListener(_handleTick);
+    _controller = config.controller ?? DefaultTabController.of(context);
+    if (_controller != null)
+      _controller.animation.addListener(_handleTick);
   }
 
-  void _updateScrollBehaviorForSelectedIndex(int selectedIndex) {
-    if (selectedIndex == 0) {
-      didUpdateScrollBehavior(scrollBehavior.updateExtents(contentExtent: 2.0, containerExtent: 1.0, scrollOffset: 0.0));
-    } else if (selectedIndex == _tabCount - 1) {
-      didUpdateScrollBehavior(scrollBehavior.updateExtents(contentExtent: 2.0, containerExtent: 1.0, scrollOffset: 1.0));
-    } else {
-      didUpdateScrollBehavior(scrollBehavior.updateExtents(contentExtent: 3.0, containerExtent: 1.0, scrollOffset: 1.0));
-    }
+  @override
+  void initState() {
+    super.initState();
+    _children = config.children;
   }
 
-  void _updateItemsAndScrollBehavior() {
-    assert(_selection != null);
-    final int selectedIndex = _selection.index;
-    assert(selectedIndex != null);
-    _updateItemsForSelectedIndex(selectedIndex);
-    _updateScrollBehaviorForSelectedIndex(selectedIndex);
+  @override
+  void dependenciesChanged() {
+    super.dependenciesChanged();
+    _initTabController();
+    _currentIndex = _controller?.index;
   }
 
   @override
-  void handleStatusChange(AnimationStatus status) {
+  void didUpdateConfig(TabBarView oldConfig) {
+    super.didUpdateConfig(oldConfig);
+    if (config.controller != oldConfig.controller)
+      _initTabController();
+    if (config.children != oldConfig.children && !_warpUnderway)
+      _children = config.children;
   }
 
   @override
-  void handleProgressChange() {
-    if (_selection == null || !_selection.valueIsChanging)
-      return;
-    // The TabBar is driving the TabBarSelection animation.
+  void dispose() {
+    _controller.animation.removeListener(_handleTick);
+    super.dispose();
+  }
 
-    final Animation<double> animation = _selection.animation;
+  void _handleTick() {
+    if (!_controller.indexIsChanging)
+      return; // This widget is driving the controller's animation.
 
-    if (animation.status == AnimationStatus.completed) {
-      _updateItemsAndScrollBehavior();
-      return;
+    if (_controller.index != _currentIndex) {
+      _currentIndex = _controller.index;
+      _warpToCurrentIndex();
     }
+  }
 
-    if (animation.status != AnimationStatus.forward)
-      return;
+  Future<Null> _warpToCurrentIndex() async {
+    if (!mounted)
```

Comment:
How can this happen?  We call this function from _handleTick and we stop listening to the animation in `dispose`.